Mathlib/Tactic/Widget/CommDiag.lean

```diff
@@ -0,0 +1,139 @@
+      e.appFn!.appFn!.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appFn!.appArg!,
+      e.appFn!.appFn!.appArg!,
+      e.appFn!.appArg!,
+      e.appArg!
+    )
+  else
+    none
+
+namespace Mathlib.Tactic.Widget
+open Lean Meta
+open ProofWidgets
+open CategoryTheory
+
+/-! ## Metaprogramming utilities for breaking down category theory expressions -/
+
+/-- Given a Hom type `α ⟶ β`, return `(α, β)`. Otherwise `none`. -/
+def homType? (e : Expr) : Option (Expr × Expr) := do
+  let some (_, _, A, B) := e.app4? ``Quiver.Hom | none
+  return (A, B)
+
+/-- Given composed homs `g ≫ h`, return `(g, h)`. Otherwise `none`. -/
+def homComp? (f : Expr) : Option (Expr × Expr) := do
+  let some (_, _, _, _, _, f, g) := f.app7? ``CategoryStruct.comp | none
+  return (f, g)
+
+/-- Expressions to display as labels in a diagram. -/
+abbrev ExprEmbeds := Array (String × Expr)
+
+/-! ## Widget for general commutative diagrams -/
+
+open scoped Jsx in
+/-- Construct a commutative diagram from a Penrose `sub`stance program and expressions `embeds` to
+display as labels in the diagram. -/
+def mkCommDiag (sub : String) (embeds : ExprEmbeds) : MetaM Html := do
+  let embeds ← embeds.mapM fun (s, h) =>
+      return (s, Html.ofTHtml <InteractiveCode fmt={← Widget.ppExprTagged h} />)
+  return Html.ofTHtml
+    <PenroseDiagram
+      embeds={embeds}
+      dsl={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"commutative.dsl"}
+      sty={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"commutativeOpt.sty"}
+      sub={sub} />
+
+/-! ## Commutative triangles -/
+
+/--
+Triangle with `homs = [f,g,h]` and `objs = [A,B,C]`
+```
+A f B
+  h g
+    C
+```
+-/
+def subTriangle := include_str ".."/".."/".."/"widget"/"src"/"penrose"/"triangle.sub"
+
+/-- Given a commutative triangle `f ≫ g = h` or `e ≡ h = f ≫ g`, return a triangle diagram.
+Otherwise `none`. -/
+def commTriangleM? (e : Expr) : MetaM (Option Html) := do
+  let e ← instantiateMVars e
+  let some (_, lhs, rhs) := e.eq? | return none
+  if let some (f, g) := homComp? lhs then
+    let some (A, C) := homType? (← inferType rhs) | return none
+    let some (_, B) := homType? (← inferType f) | return none
+    return some <| ← mkCommDiag subTriangle
+      #[("A", A), ("B", B), ("C", C),
+        ("f", f), ("g", g), ("h", rhs)]
+  let some (f, g) := homComp? rhs | return none
+  let some (A, C) := homType? (← inferType lhs) | return none
+  let some (_, B) := homType? (← inferType f) | return none
+  return some <| ← mkCommDiag subTriangle
+    #[("A", A), ("B", B), ("C", C),
+      ("f", f), ("g", g), ("h", lhs)]
+
+@[expr_presenter]
+def commutativeTrianglePresenter : ExprPresenter where
+  userName := "Commutative triangle"
+  layoutKind := .block
+  present type := do
+    if let some d ← commTriangleM? type then
+      return d
+    throwError "Couldn't find a commutative triangle."
```

Comment:
When I hit this error, I get
```
Error: Rpc error: InternalError: Couldn't find a commutative triangle.
```
Do we need the word "error" three times?

Reply:
that's probably not something that we can do anything about without changing core. This looks like the default error repr printer, applied to something like `InternalError.mk (RpcError.mk e)`.

Reply:
This is not the Lean core printer, it's something that happens in JS. I think for displaying errors it is better to err (heh) on the side of verbosity/more information. The real issue is that this error is displayed at all during normal operation.